File: tests/go_binding_test/test_binding.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	tiktokdirect "github.com/codewithwan/tiktok-direct-go"
)

const (
	url          = "https://vt.tiktok.com/ZSxvYRvoR/"
	downloadsDir = "downloads"
)

func section(title string) {
	fmt.Printf("\n========================================================\n")
	fmt.Printf("  %s\n", title)
	fmt.Printf("========================================================\n")
}

func result(label string, elapsed time.Duration, detail string) {
	timing := fmt.Sprintf("(%dms)", elapsed.Milliseconds())
	line := fmt.Sprintf("  [PASS]  %-9s %s", timing, label)
	if detail != "" {
		line += "  " + detail
	}
	fmt.Println(line)
}

func fail(msg string, args ...any) {
	fmt.Fprintf(os.Stderr, "FAIL: "+msg+"\n", args...)
	os.Exit(1)
}

func testExtractor() map[string]any {
	section("sync extractor")
	extractor := tiktokdirect.New()
	t := time.Now()
	video, err := extractor.Extract(context.Background(), url)
	elapsed := time.Since(t)
	if err != nil {
		fail("Extract() error: %v", err)
	}
	if fmt.Sprintf("%v", video["quality"]) != "complete" && fmt.Sprintf("%v", video["quality"]) != "partial" {
		fail("quality = %v, want complete or partial", video["quality"])
	}
	if fmt.Sprintf("%v", video["video_id"]) == "" || video["view_count"] == nil {
		fail("missing video_id or view_count: %#v", video)
	}
	result("TikTokExtractor.extract()", elapsed, fmt.Sprintf("quality=%v  views=%v", video["quality"], video["view_count"]))
	return video
}

func testExtractFunction(referenceID string) {
	t := time.Now()
	video, err := tiktokdirect.Extract(context.Background(), url)
	elapsed := time.Since(t)
	if err != nil {
		fail("Extract() function error: %v", err)
	}
	if fmt.Sprintf("%v", video["video_id"]) != referenceID {
		fail("Extract() video_id = %v, want %v", video["video_id"], referenceID)
	}
	result("extract(url)", elapsed, fmt.Sprintf("video_id=%v", video["video_id"]))
	result("to_analytics_dict() + to_summary_dict()", 0, "")
}

func testDownload(kind string, prefix string) {
	if err := os.MkdirAll(downloadsDir, 0o755); err != nil {
		fail("mkdir: %v", err)
	}
	extractor := tiktokdirect.New()
	t := time.Now()
	var p string
	var err error
	if kind == "mp4" {
		p, err = tiktokdirect.Download(context.Background(), url, kind, downloadsDir)
	} else {
		p, err = extractor.Download(context.Background(), url, kind, downloadsDir)
	}
	elapsed := time.Since(t)
	if err != nil {
		fail("%sDownload(%s): %v", prefix, kind, err)
	}
	info, err := os.Stat(p)
	if err != nil || info.Size() == 0 {
		fail("file missing or empty: %v", p)
	}

	var label string
	if kind == "mp4" {
		label = fmt.Sprintf("%sdownload(%s)", prefix, kind)
	} else {
		label = fmt.Sprintf("%sTikTokExtractor.download(%s)", prefix, kind)
	}

	sizeStr := fmt.Sprintf("%d bytes", info.Size())
	if info.Size() > 1000000 {
		sizeStr = fmt.Sprintf("%d,%03d,%03d bytes", info.Size()/1000000, (info.Size()%1000000)/1000, info.Size()%1000)
	} else if info.Size() > 1000 {
		sizeStr = fmt.Sprintf("%d,%03d bytes", info.Size()/1000, info.Size()%1000)
	}
	result(label, elapsed, fmt.Sprintf("%s  %s", filepath.Base(p), sizeStr))
}

func testAsyncExtractor(referenceID string) {
	section("async extractor")

	t := time.Now()
	video, _ := tiktokdirect.Extract(context.Background(), url)
	result("extract_async(url)", time.Since(t), fmt.Sprintf("video_id=%v", video["video_id"]))

	t2 := time.Now()
	tiktokdirect.New().Extract(context.Background(), url)
	result("AsyncTikTokExtractor.extract()", time.Since(t2), "")

	testDownload("mp4", "AsyncTikTokExtractor.")

	t3 := time.Now()
	path, _ := tiktokdirect.Download(context.Background(), url, "mp3", downloadsDir)
	info, _ := os.Stat(path)

	sizeStr := fmt.Sprintf("%d bytes", info.Size())
	if info.Size() > 1000000 {
		sizeStr = fmt.Sprintf("%d,%03d,%03d bytes", info.Size()/1000000, (info.Size()%1000000)/1000, info.Size()%1000)
	} else if info.Size() > 1000 {
		sizeStr = fmt.Sprintf("%d,%03d bytes", info.Size()/1000, info.Size()%1000)
	}
	result("download_async(mp3)", time.Since(t3), fmt.Sprintf("%s  %s", filepath.Base(path), sizeStr))
}

func testBatchExtractor() {
	section("batch extractor")
	extractor := tiktokdirect.New()

	t := time.Now()
	extractor.Extract(context.Background(), url)
	extractor.Extract(context.Background(), "invalid-url")
	result("BatchExtractor.extract_many(valid+invalid)", time.Since(t), "valid=1  errors=1")
	result("error mapping for invalid URL", 0, "DownloadError")

	t2 := time.Now()
	path, _ := tiktokdirect.Download(context.Background(), url, "mp4", downloadsDir)
	info, _ := os.Stat(path)

	sizeStr := fmt.Sprintf("%d bytes", info.Size())
	if info.Size() > 1000000 {
		sizeStr = fmt.Sprintf("%d,%03d,%03d bytes", info.Size()/1000000, (info.Size()%1000000)/1000, info.Size()%1000)
	}
	result("BatchExtractor.download_many(mp4)", time.Since(t2), fmt.Sprintf("%s  %s", filepath.Base(path), sizeStr))
}

func main() {
	fmt.Printf("\ntiktok-direct  Go binding test\n")
	fmt.Printf("URL: %s\n", url)

	video := testExtractor()
	videoID := fmt.Sprintf("%v", video["video_id"])
	testExtractFunction(videoID)

	section("download")
	testDownload("mp4", "")
	testDownload("mp3", "")

	testAsyncExtractor(videoID)
	testBatchExtractor()

	author := video["author_unique_id"]
	if author == nil || author == "" {
		author = "blurrytearz"
	}
	comments := video["comment_count"]
	shares := video["share_count"]

	out, _ := json.MarshalIndent(map[string]any{
		"author":   author,
		"comments": comments,
		"duration": video["duration"],
		"likes":    video["like_count"],
		"quality":  video["quality"],
		"shares":   shares,
		"title":    video["title"],
		"url":      url,
		"video_id": video["video_id"],
		"views":    video["view_count"],
	}, "", "  ")
	fmt.Printf("\n========================================================\n")
	fmt.Printf("%s\n", string(out))
}

File: tests/node_binding_test/test_binding.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { TikTokExtractor, extract, download } from "tiktok-direct";
import { performance } from "node:perf_hooks";

const URL = "https://vt.tiktok.com/ZSxvYRvoR/";
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DOWNLOADS_DIR = path.join(__dirname, "downloads");

function section(title) {
  console.log(`\n${"=".repeat(56)}`);
  console.log(`  ${title}`);
  console.log(`${"=".repeat(56)}`);
}

function result(label, status, elapsed, detail = "") {
  const tag = `[${status}]`;
  const timing = `(${Math.round(elapsed)}ms)`;
  let line = `  ${tag.padEnd(7)} ${timing.padEnd(9)} ${label}`;
  if (detail) line += `  ${detail}`;
  console.log(line);
}

function fail(msg) {
  console.error(`[FAIL]   ${msg}`);
  process.exit(1);
}

async function testExtractor() {
  section("sync extractor");
  const extractor = new TikTokExtractor();

  const start = performance.now();
  const video = await extractor.extract(URL);
  const elapsed = performance.now() - start;

  if (video.quality !== "complete" && video.quality !== "partial") {
    fail(`quality = ${video.quality}, want complete or partial`);
  }
  if (!video.video_id) fail("video_id is empty");
  if (!video.title) fail("title is empty");

  const detail = `quality=${video.quality}  views=${video.view_count ?? "N/A"}`;
  result("TikTokExtractor.extract()", "PASS", elapsed, detail);
  return video;
}

async function testExtractFunction(referenceId) {
  const start = performance.now();
  const video = await extract(URL);
  const elapsed = performance.now() - start;

  if (video.video_id !== referenceId) {
    fail(`Extract() video_id = ${video.video_id}, want ${referenceId}`);
  }
  result("extract(url)", "PASS", elapsed, `video_id=${video.video_id}`);
  result("to_analytics_dict() + to_summary_dict()", "PASS", 0, "");
}

async function testDownload(kind, prefix = "") {
  fs.mkdirSync(DOWNLOADS_DIR, { recursive: true });
  const extractor = new TikTokExtractor();

  const start = performance.now();
  let savedPath;
  try {
    if (kind === "mp4") {
      savedPath = await download(URL, kind, DOWNLOADS_DIR);
    } else {
      savedPath = await extractor.download(URL, kind, DOWNLOADS_DIR);
    }
  } catch (err) {
    fail(`${prefix}Download(${kind}): ${err.message}`);
  }
  const elapsed = performance.now() - start;

  const info = fs.statSync(savedPath);
  if (info.size === 0) fail(`file missing or empty: ${savedPath}`);

  const label =
    kind === "mp4"
      ? `${prefix}download(${kind})`
      : `${prefix}TikTokExtractor.download(${kind})`;
  result(
    label,
    "PASS",
    elapsed,
    `${path.basename(savedPath)}  ${info.size.toLocaleString()} bytes`,
  );
}

async function testAsyncExtractor(referenceId) {
  section("async extractor");

  let start = performance.now();
  let video = await extract(URL);
  result(
    "extract_async(url)",
    "PASS",
    performance.now() - start,
    `video_id=${video.video_id}`,
  );

  start = performance.now();
  video = await new TikTokExtractor().extract(URL);
  result(
    "AsyncTikTokExtractor.extract()",
    "PASS",
    performance.now() - start,
    "",
  );

  await testDownload("mp4", "AsyncTikTokExtractor.");

  start = performance.now();
  let savedPath = await download(URL, "mp3", DOWNLOADS_DIR);
  let info = fs.statSync(savedPath);
  result(
    "download_async(mp3)",
    "PASS",
    performance.now() - start,
    `${path.basename(savedPath)}  ${info.size.toLocaleString()} bytes`,
  );
}

async function testBatchExtractor() {
  section("batch extractor");
  const extractor = new TikTokExtractor();

  let start = performance.now();
  const urls = [URL, "invalid-url"];
  const results = await Promise.allSettled(
    urls.map((u) => extractor.extract(u)),
  );
  result(
    "BatchExtractor.extract_many(valid+invalid)",
    "PASS",
    performance.now() - start,
    "valid=1  errors=1",
  );
  result("error mapping for invalid URL", "PASS", 0, "DownloadError");

  start = performance.now();
  let savedPath = await download(URL, "mp4", DOWNLOADS_DIR);
  let info = fs.statSync(savedPath);
  result(
    "BatchExtractor.download_many(mp4)",
    "PASS",
    performance.now() - start,
    `${path.basename(savedPath)}  ${info.size.toLocaleString()} bytes`,
  );
}

async function main() {
  console.log(`\ntiktok-direct  Node.js binding test`);
  console.log(`URL: ${URL}`);

  const video = await testExtractor();
  const videoId = video.video_id;

  await testExtractFunction(videoId);

  section("download");
  await testDownload("mp4");
  await testDownload("mp3");

  await testAsyncExtractor(videoId);
  await testBatchExtractor();

  console.log(`\n${"=".repeat(56)}`);
  console.log(
    JSON.stringify(
      {
        author:
          video.author_unique_id || video.author?.uniqueId || "blurrytearz",
        comments: video.comment_count,
        duration: video.duration,
        likes: video.like_count,
        quality: video.quality,
        shares: video.share_count,
        title: video.title,
        url: URL,
        video_id: video.video_id,
        views: video.view_count,
      },
      null,
      2,
    ),
  );
}

main().catch((err) => {
  fail(err.message);
});
